From 58eeabf6a0416a4f2cc4d58fcc2242d3aef80205 Mon Sep 17 00:00:00 2001
From: Giuseppe D'Angelo <giuseppe.dangelo@kdab.com>
Date: Mon, 05 Oct 2020 10:33:21 +0200
Subject: [PATCH] xdgdesktopportal: properly fix a dangling pointer

The previous fix didn't actually work, as it kept a reference
into a container which could have been muted by the time the
reference was going to be used.

Use an index instead.

Amends 32c09ea5b0f529418eece63de5c3b2c206f62896.

Change-Id: Ib855b4a663c281467e46536b98a0ce2b961f19ee
Pick-to: 5.15
Task-number: QTBUG-87143
---

diff --git a/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportalfiledialog.cpp b/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportalfiledialog.cpp
index ae2532e..cf3c2c9 100644
--- a/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportalfiledialog.cpp
+++ b/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportalfiledialog.cpp
@@ -210,7 +210,7 @@
     qDBusRegisterMetaType<FilterList>();
 
     FilterList filterList;
-    Filter* selectedFilter = nullptr;
+    auto selectedFilterIndex = filterList.size() - 1;
 
     d->userVisibleToNameFilter.clear();
 
@@ -236,7 +236,7 @@
             filterList << filter;
 
             if (!d->selectedMimeTypeFilter.isEmpty() && d->selectedMimeTypeFilter == mimeTypefilter)
-                selectedFilter = &filterList.last();
+                selectedFilterIndex = filterList.size() - 1;
         }
     } else if (!d->nameFilters.isEmpty()) {
         for (const QString &nameFilter : d->nameFilters) {
@@ -265,7 +265,7 @@
                 d->userVisibleToNameFilter.insert(userVisibleName, nameFilter);
 
                 if (!d->selectedNameFilter.isEmpty() && d->selectedNameFilter == nameFilter)
-                    selectedFilter = &filterList.last();
+                    selectedFilterIndex = filterList.size() - 1;
             }
         }
     }
@@ -273,9 +273,8 @@
     if (!filterList.isEmpty())
         options.insert(QLatin1String("filters"), QVariant::fromValue(filterList));
 
-    if (selectedFilter) {
-        options.insert(QLatin1String("current_filter"), QVariant::fromValue(*selectedFilter));
-    }
+    if (selectedFilterIndex != -1)
+        options.insert(QLatin1String("current_filter"), QVariant::fromValue(filterList[selectedFilterIndex]));
 
     options.insert(QLatin1String("handle_token"), QStringLiteral("qt%1").arg(QRandomGenerator::global()->generate()));
 
